rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -1,226 +1,370 @@
     """
     training = False
-    if not zero_state_features:
+    if not fake_features:
         training = True
-        zero_state_features = defaultdict(list)
+        fake_features = defaultdict(list)
 
     # unify format of incoming features
-    if isinstance(tracker_state_features[0], Dict):
-        tracker_state_features = [[dicts] for dicts in tracker_state_features]
+    if isinstance(features[0], Dict):
+        features = [[dicts] for dicts in features]
 
-    state_to_tracker_features = surface_attributes(tracker_state_features)
+    attribute_to_features = _surface_attributes(features, featurizers)
 
     attribute_data = {}
 
     # During prediction we need to iterate over the zero features attributes to
     # have all keys in the resulting model data
     if training:
-        attributes = list(state_to_tracker_features.keys())
+        attributes = list(attribute_to_features.keys())
     else:
-        attributes = list(zero_state_features.keys())
+        attributes = list(fake_features.keys())
 
     # In case an attribute is not present during prediction, replace it with
     # None values that will then be replaced by zero features
     dialogue_length = 1
-    for tracker_features in state_to_tracker_features.values():
-        dialogue_length = max(dialogue_length, len(tracker_features[0]))
-    empty_features = [[None] * dialogue_length]
+    num_examples = 1
+    for _features in attribute_to_features.values():
+        num_examples = max(num_examples, len(_features))
+        dialogue_length = max(dialogue_length, len(_features[0]))
+    empty_features = [[None] * dialogue_length] * num_examples
 
     for attribute in attributes:
         attribute_data[attribute] = _features_for_attribute(
             attribute,
             empty_features,
-            state_to_tracker_features,
+            attribute_to_features,
             training,
-            zero_state_features,
+            fake_features,
+            consider_dialogue_dimension,
         )
 
     # ensure that all attributes are in the same order
     attribute_data = OrderedDict(sorted(attribute_data.items()))
 
-    return attribute_data, zero_state_features
+    return attribute_data, fake_features
 
 
 def _features_for_attribute(
     attribute: Text,
     empty_features: List[Any],
-    state_to_tracker_features: Dict[Text, List[List[List["Features"]]]],
+    attribute_to_features: Dict[Text, List[List[List["Features"]]]],
     training: bool,
-    zero_state_features: Dict[Text, List["Features"]],
-) -> Dict[Text, List[np.ndarray]]:
-    """Create the features for the given attribute from the tracker features.
+    fake_features: Dict[Text, List["Features"]],
+    consider_dialogue_dimension: bool,
+) -> Dict[Text, List[FeatureArray]]:
+    """Create the features for the given attribute from the all examples features.
 
     Args:
         attribute: the attribute
         empty_features: empty features

```

Comment:
The explanation is not really helpful here 😅 
```suggestion
        attribute: attribute of Message to be featurized
        empty_features: default features for examples with missing attribute value
```

Reply:
they're not really default features. they are literally empty features

Reply:
So the difference between `fake_features` and `empty_features` is that `fake_features` is a vector of 0s and `empty_features` is a list of Nones?

Reply:
actually, they are the same thing. I think ...

Reply:
they're literally empty. e.g. arrays with first dim of `0`, but they're np.zeros arrays

Reply:
@Ghostvv  could you please take this over

Reply:
> So the difference between fake_features and empty_features is that fake_features is a vector of 0s and empty_features is a list of Nones?

yes, but only, fake_features already contain all correct dimensions

Reply:
renamed `empty_features` to `absent_features`